### PR TITLE
[ts-command-line] Update to `argparse@2.0.1` and default `allow_abbrev` to false

### PR DESCRIPTION
**Pull request by D4N14L**
## Summary

This PR bumps the version of argparse consumed to `2.0.1`. It also defaults `allow_abbrev` to `false` when creating the underlying argument parser, to avoid issues with unexplicit parameter usage (ex. `--debug` matching `--debug-heft-reporter` which comes from the `@rushstack/heft-jest-plugin` package). More documentation on the parameter can be found here: https://docs.python.org/3/library/argparse.html#allow-abbrev

## Details

API usage across the entire argparse surface had to be updated, since the underlying library now uses `snake_casing`, but behaviour remains the same.

## How it was tested

All unit tests passed for ts-command-line, new test added for abbreviation scenario and confirmed the test failed without setting `allow_abbrev` to `false`.

## Impacted documentation

None. This was undocumented and unintended behaviour.

### CODE REVIEW

File: libraries/ts-command-line/src/providers/CommandLineParameterProvider.ts [hunk 1]
```diff
@@ -690,21 +683,50 @@ export abstract class CommandLineParameterProvider {
...
+        break;
+      default:
+        // Should never happen, but guard against it just in case. Using a normal error instead
+        // of an InternalError since this package intentionally avoids using node-core-library.
+        throw new Error(`Unexpected number of parameter names provided: ${names.length}`);
+    }
 
-    if (parameter.undocumentedSynonyms?.length) {
-      argumentGroup.addArgument(parameter.undocumentedSynonyms, {
+    for (const undocumentedSynonym of parameter.undocumentedSynonyms || []) {
```
Comment (dmichon-msft) [inline libraries/ts-command-line/src/providers/CommandLineParameterProvider.ts:726]: Wrapping in the if is marginally more efficient

File: libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap [hunk 2]
```diff
@@ -550,55 +548,54 @@ exports[`CommandLineParameter prints the action help 1`] = `
                           [--string-with-default TEXT]
                           [--string-with-undocumented-synonym TEXT]
                           [-l LIST_ITEM]
-                          
 
 a longer description
 
-Optional arguments:
-  -h, --help            Show this help message and exit.
+optional arguments:
```
Comment (dmichon-msft) [inline libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap:554]: This inconsistent formatting looks bad; if the base is changing should we change to match?

**Review by dmichon-msft — APPROVED** [review]

File: libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap (hunk 2 above)
Comment (octogonz) [reply libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap:554]: I don't think so. Beginning sentences with a lower case letter looks objectively worse.  I checked a few other popular CLI tools (`tsc`, `eslint`, `PowerShell`, `gcc`) and their `--help` always uses proper English capitalization. Doing so also encourages grammatically complete sentences, which may inspire better documentation. Notably `python --help` does show unprofessional lowercase capitalization, however Python's own `pip --help` uses caps like all the other examples.

I believe this behavior is customizable by implementing our own `argparse` help formatter.

**Comment by D4N14L** [comment]
Closed in favor of #4405 which implements the main goal of this PR, without wholesale disabling `allow_abbrev` or needing to bump argparse

File: libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap (hunk 2 above)
Comment (octogonz) [reply libraries/ts-command-line/src/test/__snapshots__/CommandLineParameter.test.ts.snap:554]: The API allows us to suppress the default `--help` parameter and define it ourselves.

I created an issue https://github.com/nodeca/argparse/issues/175 requesting a more general solution, however the NPM package hasn't been updated in 3 years, so we'll see if they reply. 😄

**Pull request closed without merging**